### ISSUE TITLE
[Tune] HyperBand and ASHA not scheduling trials correctly.

### ISSUE THREAD

**Issue by mvanness354** [ray-project/ray#37605]
### What happened + What you expected to happen

I've been trying to run HyperBand (and ASHA) with XGBoost, and have noticed that the trials are not being trained how I would expect them to be trained. For example, consider a hyperparameter optimization run which starts with 8 trials, where the maximum number of iterations is 8. I would expect HyperBand to run each trial for 1 iteration (regardless of the number of concurrent trials), and then terminate half of the trials, then run the remaining 4 trials for 2 more iteration, terminate half the trials, etc. However, I have found that the Ray HyperBand scheduler used with XGBoost will instead run the first 4 trials for all 8 iterations, run the second set of 4 trials, and then pause and never actually finish.

Here is a [Google Colab notebook](https://colab.research.google.com/drive/1oZ-dkwuidBG3AzqtL-8w5SQdmUYAExSD?usp=sharing) that fully reproduces the issue.

### Versions / Dependencies

`python==3.10.6`
`ray==2.5.1`
`xgboost_ray==0.1.16`

### Reproduction script

```
# Also see Colab link in description

from sklearn.datasets import load_breast_cancer
from ray.data import from_pandas
from ray.train.xgboost import XGBoostTrainer
from ray.tune.tuner import Tuner, TuneConfig
from ray import air, tune
from ray.air import session
from ray.tune.schedulers import AsyncHyperBandScheduler, HyperBandScheduler
from ray.tune import CLIReporter

data = load_breast_cancer(as_frame=True)
data_set = data["data"]
data_set["target"] = data["target"]
train_set = from_pandas(data_set)

xgb_params = {
    "objective": "binary:logistic",
    "eval_metric": ["logloss", "error"],
}

trainer = XGBoostTrainer(
    label_column="target",
    num_boost_round=8,
    datasets={"train": train_set},
    params=xgb_params
)

param_space = {
    "params": {
        "max_depth": tune.randint(5, 15),
        "eta": tune.loguniform(1e-5, 1),
    }
}

# scheduler = AsyncHyperBandScheduler(grace_period=1, max_t=8, reduction_factor=2)
scheduler = HyperBandScheduler(max_t=8, reduction_factor=2)

metric = "train-logloss"

reporter = CLIReporter(print_intermediate_tables=True)
tuner = Tuner(
    trainer,
    param_space=param_space,
    tune_config=TuneConfig(
        num_samples=8,
        metric=metric,
        scheduler=scheduler,
        mode="min",
    ),
    run_config=air.RunConfig(
        progress_reporter=reporter,
    )
)


bst = tuner.fit()
```

### Issue Severity

High: It blocks me from completing my task.

**Comment by xwjiang2010** [comment]
I haven't verified the exact complaint but running the script does look weird. Especially, 6 of the 8 trials are finished while the rest 2 are stuck at PENDING state although no resources are being used in the whole cluster.
Also doesn't seem to be related to the new execution logic, as the same thing happened with `TUNE_NEW_EXECUTION=0`

**Comment by krfricke** [comment]
Confirmed that this is an issue, looking into it today

**Comment by mvanness354** [comment]
Hi @krfricke, thanks for taking a look at this. I do want to add, though, that since I made this issue I spent a bit of time working on a new implementation of ASHA that I posted in [this issue](https://github.com/ray-project/ray/issues/32634#issuecomment-1689018637) that more closely matches the original ASHA paper and also better suited my needs. After making that implementation, I believe that the issue that causes the current HyperBand and ASHA implementations to not work in the colab notebook I shared is that these implementations seem to only make sense when there are enough resources to start many (if not all) jobs in parallel, which is not possible with only 1 CPU in colab.
